### PR TITLE
feat(chain)!: game shard query precompile

### PR DESCRIPTION
**Pull request by technicallyty**
## What is the purpose of the change

adds a new precompile method to `router` called `query`, which allows EVM smart contracts to make query's against Cardinal.

## Brief Changelog

- updates `Read` interface with methods that add EVM support
- updates `NewReadType` function - adds parameter `supportEVM bool` which, when true, will dynamically generate the ABI equivalent types of the request and reply types.
- adds `query` method to `router` precompile contract
- extends `Router` proto to have `query` rpc method
- implements query method client and server implementations

## Testing and Verifying
pending e2e tests
`cardinal/ecs/tests/read_test.go`

### CODE REVIEW

File: cardinal/ecs/abi.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	"uint":           "uint64",
+	"int":            "int64",
+	"common.Address": "address",
+	"uint8":          "uint8",
+	"uint16":         "uint16",
+	"uint32":         "uint32",
+	"uint64":         "uint64",
+}
+
+func GenerateABIType(goStruct any) (*abi.Type, error) {
```
Comment (jerargus) [inline cardinal/ecs/abi.go:23]: This seems like a good candidate for a table driven test.

**Review by jerargus — CHANGES_REQUESTED** [review]

File: cardinal/ecs/abi.go (hunk 1 above)
Comment (technicallyty) [reply cardinal/ecs/abi.go:25]: i added more test coverage, but seemed more straightforward to just have a massive struct with all supported types to make sure we can pack & unpack, and then individual tests for edge cases. 

if you still think its valuable to have these be table driven, lmk, i can look into this more

**Review by jerargus — APPROVED** [review]

File: cardinal/evm/server.go [hunk 2]
```diff
@@ -107,6 +110,26 @@ func (s *srv) SendMsg(ctx context.Context, msg *routerv1.MsgSend) (*routerv1.Msg
 		return nil, err
 	}
 	// add transaction to the world queue
-	s.txh.AddTransaction(itx.ID(), tx, nil)
+	s.world.AddTransaction(itx.ID(), tx, nil)
 	return &routerv1.MsgSendResponse{}, nil
 }
+
+func (s *srv) QueryShard(ctx context.Context, req *routerv1.QueryShardRequest) (*routerv1.QueryShardResponse, error) {
```
Comment (smsunarto) [inline cardinal/evm/server.go:117]: Im cautious of using the word "query" here since it means something in ECS. Maybe we should rename this to read for consistency?

File: chain/precompile/router/router.go [hunk 3]
```diff
@@ -39,3 +39,11 @@ func (c *Contract) Send(
 	_, err := c.rtr.Send(ctx, namespace, pCtx.MsgSender().String(), messageID, message)
 	return nil, err
 }
+
+func (c *Contract) Query(
```
Comment (smsunarto) [inline chain/precompile/router/router.go:43]: Read?

File: chain/precompile/router/router.go [hunk 4]
```diff
@@ -39,3 +39,11 @@ func (c *Contract) Send(
 	_, err := c.rtr.Send(ctx, namespace, pCtx.MsgSender().String(), messageID, message)
 	return nil, err
 }
```
Comment (smsunarto) [inline chain/precompile/router/router.go:41]: We should probably make it clearer here `sendMessage` sounds about right

File: chain/contracts/src/cosmos/precompile/router.sol [hunk 5]
```diff
@@ -1,5 +1,10 @@
 pragma solidity ^0.8.4;
 
 interface IRouter {
-    function send(bytes calldata message, uint64 messageID, string calldata namespace) external returns (bytes memory);
+    function send(bytes calldata message, uint64 messageID, string calldata namespace)
+        external
+        returns (bytes memory);
+    function query(bytes calldata request, string calldata resource, string calldata namespace)
```
Comment (namrapatel) [inline chain/contracts/src/cosmos/precompile/router.sol:7]: if we change from query -> read prob also gotta change here

File: cardinal/ecs/read.go [hunk 6]
```diff
@@ -57,3 +118,73 @@ func (r *ReadType[req, rep]) HandleRead(w *World, bz []byte) ([]byte, error) {
 	}
 	return bz, nil
 }
+
+func (r *ReadType[req, rep]) DecodeEVMRequest(bz []byte) (any, error) {
+	if r.requestABI == nil {
+		return nil, errors.New("cannot call DecodeEVMRequest without using the WithEVMSupport option when " +
+			"creating the read")
```
Comment (namrapatel) [inline cardinal/ecs/read.go:125]: just wondering: are we using this + in the error for readability?

File: chain/precompile/router/router.go (hunk 4 above)
Comment (technicallyty) [reply chain/precompile/router/router.go:41]: not entirely sure whats being said here

File: cardinal/ecs/read.go (hunk 6 above)
Comment (technicallyty) [reply cardinal/ecs/read.go:125]: yeah, linter has line length rules

File: cardinal/evm/server.go (hunk 2 above)
Comment (technicallyty) [reply cardinal/evm/server.go:117]: yeah, i was veering either way, cause `Read` is quite specific naming to Cardinal, and the precompile is eventually generalized to many different game shards, where i imagine query will be the default word for these types of things. 

the more i think about this, the more i want to rename our read system to query honestly lol

File: cardinal/evm/server.go [hunk 7]
```diff
@@ -21,33 +20,37 @@ var (
...
-
-// TxHandler is a type that gives access to transaction data in the ecs.World, as well as access to queue transactions.
-type TxHandler interface {
-	AddTransaction(transaction.TypeID, any, *sign.SignedPayload) uint64
-	ListTransactions() ([]transaction.ITransaction, error)
-}
+// IReadTypes maps read resource names to the underlying IRead
+type IReadTypes map[string]ecs.IRead
 
 type srv struct {
```
Comment (namrapatel) [inline cardinal/evm/server.go:29]: not to be a troll, but feels like someone reading this code would be confused by what this is for cuz of the name? feel free to resolve/ignore if you think it doesn't matter much or there isn't a better name

Comment (technicallyty) [reply cardinal/evm/server.go:29]: glad you called this out, cause these types shouldn't be exported either, since they're only used within this package. 

I've renamed them to describe what they're actually doing, i.e. mapping some identifier to some type - txById and readByName

File: chain/precompile/router/router.go (hunk 4 above)
Comment (smsunarto) [reply chain/precompile/router/router.go:41]: yeah

File: cardinal/evm/server.go (hunk 2 above)
Comment (smsunarto) [reply cardinal/evm/server.go:117]: grrr, yeah i guess query makes more sense

File: cardinal/evm/server.go (hunk 7 above)
Comment (technicallyty) [reply cardinal/evm/server.go]: re: `srv` updated to be `msgServerImpl`

**Review by namrapatel — APPROVED** [review]

File: chain/precompile/router/router.go (hunk 4 above)
Comment (technicallyty) [reply chain/precompile/router/router.go:41]: updooted [refactor: send -> SendMessage](https://github.com/Argus-Labs/world-engine/pull/186/commits/a2c7e08701ffed26bbfec8453262e40655a3b9df)
